pkg/reconciler/buildrun/runner/taskrun.go

```diff
@@ -0,0 +1,491 @@
+		return false
+	}
+	return succeededCondition.IsFalse() || succeededCondition.IsTrue()
+}
+
+func (t *TaskRunBuildRunner) Validate(ctx context.Context, client client.Client) *BuildRunnerValidationError {
+	err := resources.CheckTaskRunVolumesExist(ctx, client, t.TaskRun)
+	if err == nil {
+		return nil
+	}
+	validationErr := &BuildRunnerValidationError{
+		Message: err.Error(),
+	}
+	if apierrors.IsNotFound(err) {
+		validationErr.Terminal = true
+		validationErr.ReasonCode = string(v1beta1.VolumeDoesNotExist)
+	}
+	return validationErr
+}
+
+func (t *TaskRunBuildRunner) Cancel(ctx context.Context, client client.Client) error {
+	// patch tekton taskrun a la tkn to start tekton's cancelling logic
+	trueParam := true
+	return t.patchTaskRun(ctx, client, t.TaskRun, "replace", "/spec/status", pipelineapi.TaskRunSpecStatusCancelled, metav1.PatchOptions{Force: &trueParam})
+}
+
+type patchStringValue struct {
+	Op    string `json:"op"`
+	Path  string `json:"path"`
+	Value string `json:"value"`
+}
+
+func (t *TaskRunBuildRunner) patchTaskRun(ctx context.Context, k8sClient client.Client, tr *pipelineapi.TaskRun, op, path, value string, opts metav1.PatchOptions) error {
+	payload := []patchStringValue{{
+		Op:    op,
+		Path:  path,
+		Value: value,
+	}}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+	patch := client.RawPatch(types.JSONPatchType, data)
+	patchOpt := client.PatchOptions{Raw: &opts}
+	return k8sClient.Patch(ctx, tr, patch, &patchOpt)
+}
+
+func (t *TaskRunBuildRunner) SyncBuildRunStatus(ctx context.Context, client client.Client, buildRun *v1beta1.BuildRun) error {
+
+	if buildRun.Status.StartTime == nil && t.GetStartTime() != nil {
+		buildRun.Status.StartTime = t.GetStartTime()
+	}
+
+	if buildRun.Status.CompletionTime == nil && t.GetCompletionTime() != nil {
+		buildRun.Status.CompletionTime = t.GetCompletionTime()
+	}
+
+	if len(t.Status.Results) > 0 {
+		// TODO: this logic could be optimized so we iterate through the result values only once.
+		updateBuildRunStatusWithSourceResult(buildRun, t.Status.Results)
+		updateBuildRunStatusWithOutputResult(buildRun, t.Status.Results)
+	}
+
+	succeededCondition := t.Status.GetCondition(apis.ConditionSucceeded)
+	if succeededCondition == nil {
+		return nil
+	}
+
+	if err := t.updateBuildRunUsingTaskRunCondition(ctx, client, buildRun, succeededCondition); err != nil {
+		return err
+	}
+
+	t.updateBuildRunUsingTaskFailures(ctx, client, buildRun)
+
+	return nil
+}
+
+const (
+	defaultSourceName          = "default"
+	sourceTimestampName        = "source-timestamp"
+	prefixParamsResultsVolumes = "shp"
+	imageDigestResult          = "image-digest"
+	imageSizeResult            = "image-size"
+	imageVulnerabilities       = "image-vulnerabilities"
+)
+
+func updateBuildRunStatusWithSourceResult(buildrun *v1beta1.BuildRun, results []pipelineapi.TaskRunResult) {
+	buildSpec := buildrun.Status.BuildSpec
+
+	if buildSpec.Source == nil {
+		return
+	}
```

Comment:
This is not accurate. `status.buildSpec` is a copy of the Build spec. But the real source may be the BuildRun source. For example, in case of a local source, this causes an overwrite.